### ISSUE TITLE
JCache `CacheLoader` required to be `Serializable`

### ISSUE THREAD

**Issue by dansiviter** [hazelcast/hazelcast#6465]
Although the [example](http://docs.hazelcast.org/docs/3.5/manual/html-single/hazelcast-documentation.html#cacheloader) alludes to it there is no explicit mention to why a `javax.cache.integration.CacheLoader` must implement `java.io.Serializable`. If either it, or it's  dependants, do not implement it will result in a `com.hazelcast.nio.serialization.HazelcastSerializationException`. Why is there a requirement to serialise a `CacheLoader`?

**Comment by serkan-ozal** [comment]
Hi @dansiviter 
You are right, in fact there is no need for `CacheLoader` to implement `Serializable`.  `CacheLoader` are created with factory so its factory must be serializable.

To create factory, there is `javax.cache.configuration.FactoryBuilder` and it has some methods to create factory for `CacheLoader`:
* `public static <T> Factory<T> factoryOf(Class<T> clazz)`: In this usage, `CacheLoader` doesn't need to implement `Serializable` since its instance is not serialized in distributed environment inside cache configuration but factory will be serialized so `javax.cache.configuration.Factory` implements `Serializable`.
* `public static <T extends Serializable> Factory<T> factoryOf(T instance)`: In this usage, since `CacheLoader` instance is serialized inside factory, it must be `Serializable`.

So as a result, implementing `Serializable` is not necessary for `CacheLoader` and it depends the how it is configured.

**Comment by dansiviter** [comment]
This is where is starts to get confusing as the class I'm using has a requirement on other classes which are Spring managed. I can't serialise the whole `ApplicatonContext` so I'm unsure how to inject the classes I need if the factory is created by `javax.cache.configuration.FactoryBuilder.ClassFactory`. In the [example](http://docs.hazelcast.org/docs/latest/manual/html/jcache-apicacheloader.html) the `UserDao`, or all it's member variables, surely can't be serialised too? If it's reliant on something like `javax.persistence.EntityManager` (which doesn't implement `Serializable`) then this will fail too. I know I can make the value `transient` but this doesn't address injecting in the `EntityManager`.

**Comment by serkan-ozal** [comment]
@dansiviter I couldn't understand your issue/requirement. Can you explain it?

**Comment by dansiviter** [comment]
So if I was using IoC I would want to inject a persistence context into a `CacheLoader` e.g.:

```
public void UserLoader implement CacheLoader<String, User> {
  @PersistenceContext
  private EntityManager entityManager;

  @Override
  public User load(String key) {
    return this.entityManager.find(User.class, key);
  }
  ...
}
```

To inject `javax.persistence.EntityManager` (which does not extend `Serializable`) either Hazelcast would have to do resolve and inject via `factoryOf(Class<T> clazz)` or we would have to pass a concrete value which was already resolved using `factoryOf(T instance)`. Unfortunately, I don't believe Hazelcast will inject it so our only choice is using a concrete value which will fail due to `EntityManager` not being serialisable.

Basically, it's a catch 22 situation and no clear way to be able to use JPA or a DAO style loading as a read-through cache.

**Comment by dansiviter** [comment]
Any update? As I see it there is no practical way to use Hazelcast JCache in read-through mode using `PersistenceContext` or any other form of injection.

**Comment by serkan-ozal** [comment]
Hi @dansiviter , 
I think it is not JCache spec or Hazelcast issue since it depends on your `CacheLoaderFactory` implementation for how to create `CacheLoader`. 

I am not Spring expert but I think there can be some workarounds to inject dependencies into `CacheLoader` inside `CacheLoaderFactory`. 

For example,

``` java
@Component
public class SpringUtil implements ApplicationContextAware {

	private static ApplicationContext applicationContext;
	
	public static ApplicationContext getApplicationContext() {
		return applicationContext;
	}
	
	@Override
	public void setApplicationContext(ApplicationContext context) throws BeansException {
		applicationContext = context;
	}

	public static <T> T injectDependencies(T bean) {
		applicationContext.getAutowireCapableBeanFactory().autowireBean(bean);
		return bean;
        }

}
```

This example is not a perfect/optimum solution but I think it can help you.

**Comment by dansiviter** [comment]
Thanks for your assistance. I'll give it a go and report back. 

It does surprise me there isn't more support for CDI and Spring though.

**Comment by dansiviter** [comment]
Looking closer at your suggestion is means looking up `ApplicationContext` `static`ally and that's only going to be painful.

Therefore, I feel it would be a reasonable suggestion to request Hazelcast can perform CDI from within `CacheLoader`, `CacheWriter`, etc., as suggested by the JCache specification:

> In implementations and deployment environments that support it, customizations may additionally
utilize technologies such as resource injection (e.g: CDI) to allow direct access to application and
deployment specific resources. There is no requirement however that implementations support this
ability

Therefore I've added two new items to add support for CDI (#6617) and Spring (#6616) within the cache read/write mechanisms and potentially in the distributed tasks.
